diff --git a/dist/index.js b/dist/index.js
index 8b8e0d078e27474da1cf58ce3fef1d7acefb1cd4..0753137ff48b7d34338db98c3568c82e9798b3b7 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -146,9 +146,10 @@ async function removeTypesFromVueSfcScript(code, fileName, script, templateAst,
     (0, import_template_ast_types.traverse)(templateAst, {
       enter(node) {
         if ((0, import_template_ast_types.isSimpleExpressionNode)(node) && !node.isStatic) {
-          expressions.add(node.content);
+          const content = node.content.trim();
+          expressions.add(`[${content}]`);
         } else if ((0, import_template_ast_types.isComponentNode)(node)) {
-          expressions.add(node.tag);
+          expressions.add(`[${node.tag}]`);
         }
       }
     });