// @morev/vue-transitions@2.3.6/nuxt/module.cjs
'use strict';

const fs = require('fs');
const utils = require('@morev/utils');
const kit = require('@nuxt/kit');

const ALL_COMPONENTS = {
  TransitionExpand: "TransitionExpand",
  TransitionSlide: "TransitionSlide",
  TransitionScale: "TransitionScale",
  TransitionFade: "TransitionFade"
};
const module$1 = kit.defineNuxtModule({
  meta: {
    name: "@morev/vue-transitions/nuxt",
    configKey: "vueTransitions"
  },
  defaults: {},
  async setup(options, nuxt) {
    var _a;
    const DIRECTORY_NAME = "vue-transitions";
    const resolver = kit.createResolver((typeof document === 'undefined' ? new (require('u' + 'rl').URL)('file:' + __filename).href : (document.currentScript && document.currentScript.src || new URL('module.cjs', document.baseURI).href)));
    const entries = options.components ?? ALL_COMPONENTS;
    if (!Object.values(entries).filter(Boolean).length)
      return;
    (_a = nuxt.options).css ?? (_a.css = []);
    nuxt.options.css.push(`@morev/vue-transitions/styles`);
    nuxt.options.build.transpile.push("@morev/utils");
    const templateContents = fs.readFileSync(resolver.resolve("template.vue"), { encoding: "utf8" });
    const componentsDir = resolver.resolve(DIRECTORY_NAME);
    try {
      fs.existsSync(componentsDir) && fs.unlinkSync(componentsDir);
    } catch {
    }
    fs.mkdirSync(componentsDir, { recursive: true });
    utils.tsObject.entries(entries).forEach(([originalPascalName, neededName]) => {
      const customProps = utils.mergeObjects(
        options.defaultProps ?? {},
        options.componentDefaultProps?.[originalPascalName]
      );
      const neededKebabName = utils.kebabCase(neededName);
      const propsDeclaration = utils.isEmpty(customProps) ? "$attrs" : JSON.stringify(customProps).replace(/}$/, ",...$$attrs}").replace(/"/g, "'");
      fs.writeFileSync(
        resolver.resolve(`${DIRECTORY_NAME}/${neededKebabName}.vue`),
        templateContents.replace(/<%= options\.propsDeclaration %>/g, propsDeclaration).replace(/<%= options\.originalPascalName %>/g, originalPascalName).replace(/<%= options\.neededName %>/g, neededName)
      );
    });
    kit.addComponentsDir({
      path: componentsDir,
      pathPrefix: false,
      watch: false
    });
    const typeMappings = Object.entries(entries).reduce((acc, [originalName, neededName]) => {
      acc.push(`${neededName}: DefineComponent<ComponentPropsAndEmits['${originalName}']>;`);
      return acc;
    }, []).join("\n		");
    const typesPath = kit.addTemplate({
      filename: `types/vue-transitions.d.ts`,
      src: resolver.resolve("template.d.ts"),
      write: true,
      options: {
        declarations: `
		${typeMappings}
	`
      }
    }).dst;
    nuxt.hook("prepare:types", (types) => {
      types.references.push({ path: typesPath });
    });
  }
});

module.exports = module$1;
